smac_planner/include/smac_planner/costmap_downsampler.hpp

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2020, Carlos Luis
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License. Reserved.
+
+#ifndef SMAC_PLANNER__DOWNSAMPLER_HPP_
+#define SMAC_PLANNER__DOWNSAMPLER_HPP_
+
+#include <algorithm>
+#include "nav2_costmap_2d/costmap_2d_ros.hpp"
+#include "smac_planner/constants.hpp"
+
+namespace smac_planner
+{
+
+/**
+ * @class smac_planner::CostmapDownsampler
+ * @brief A costmap downsampler for more efficient path planning
+ */
+class CostmapDownsampler
+{
+public:
+  /**
+   * @brief A constructor for CostmapDownsampler
+   */
+  explicit CostmapDownsampler()
+  : _downsampling_factor(0)
+  {
+  }
+
+  /**
+   * @brief A destructor for CostmapDownsampler
+   */
+  ~CostmapDownsampler()
+  {
+  }
+
+  /**
+   * @brief Initialize the downsampled costmap object
+   * @param costmap The costmap we want to downsample
+   * @param downsampling_factor Multiplier for the costmap resolution
+   * @return A shared ptr to the downsampled costmap
+   */
+  std::shared_ptr<nav2_costmap_2d::Costmap2D> initialize(
+    nav2_costmap_2d::Costmap2D * costmap,
+    const unsigned int & downsampling_factor)
+  {
+    _size_x = costmap->getSizeInCellsX();
+    _size_y = costmap->getSizeInCellsY();
+    _costmap = costmap;
+    _downsampling_factor = downsampling_factor;
+    float rounded_resolution = _downsampling_factor * costmap->getResolution();
+    unsigned int map_cells_size_x = ceil(static_cast<float>(_size_x) / _downsampling_factor);
+    unsigned int map_cells_size_y = ceil(static_cast<float>(_size_y) / _downsampling_factor);
+
+    _downsampled_costmap = std::make_shared<nav2_costmap_2d::Costmap2D>
+            (map_cells_size_x, map_cells_size_y, rounded_resolution,
+             _costmap->getOriginX(), _costmap->getOriginY(), UNKNOWN);
+    return _downsampled_costmap;
```

Comment:
I'm not totally following the logic behind returning the shared pointer in the initialization function. Why not return it in the downsample() method? I think logically that would make more sense (`downsampled_costmap = downsampler->downsample()`), is there some memory / copy reason for this?

Reply:
The reason is I need to pass a Costmap2D ptr (with an already initialized Costmap2D) to the constructor of the costmap publisher. That costmap publisher constructor would throw an error if the pointer has not been initialized.